### ISSUE TITLE
Dependency shading is not complete, or is not configured correctly

### ISSUE THREAD

**Issue by jdcaperon** [snowflakedb/snowflake-ingest-java#356]
When adding the shaded version of the snowflake dependency to our closure, we noticed that it appears that shading is not configured correctly as there are a large number of conflicts in our dependency closure. It appears that shading is configured correctly for some dependencies and not for others, particularly some transitive dependencies are not covered it seems.

I couldn't find why shading was introduced in the first place, there's very little information other than a Relocate all dependencies to internal to solve dependency conflict problem comment: https://github.com/snowflakedb/snowflake-ingest-java/pull/19

In the current 1.1.0 shaded jar there are quite a few packages that conflict with the ones in our dependency closure, including com.google.common, com.codahale.metrics, and org.eclipse.jetty

For now we are going to switch to the unshaded dependency to resolve these issues but it would be good to reconfigure the shading to be correct for the shaded version of the dependency.

**Comment by leszekgruchala** [comment]
We have faced the issue too. Our solution was to change the order of loaded jars, to make snowflake-sdk the last one.

**Comment by sfc-gh-tzhang** [comment]
@sfc-gh-tjones @sfc-gh-dwang Any idea about the latest question? I think the protobuf library is part of the hadoop-common? But I'm not sure why they didn't hit this issue before the refactoring.

**Comment by sfc-gh-dwang** [comment]
@sfc-gh-tzhang com.google.protobuf:protobuf-java 2.5 is brought by hadoop-common 3.3.5, likely 3.3.4 uses some old version of protobuf

**Comment by leszekgruchala** [comment]
@sfc-gh-tzhang @sfc-gh-tjones @sfc-gh-dwang Can you please explain why you keep the same packages of the shaded libraries? They should be different, because now the shaded jars are still available on the classpath with the original packages and are conflicting with others.

As I understand, you bring an uber jar with all your dependencies still available in the client's classpaths even though you try to hide them. It's worse than if you provide them as regular dependencies defined in the pom.xml, because now the client's can't properly evict the incorrect library versions.

I don't know about Maven, but in SBT there is a possibility to rename the packages
https://stackoverflow.com/a/40404371/804812

```
assemblyShadeRules in assembly ++= Seq(
  ShadeRule.rename("com.typesafe.config.**" -> "my_conf.@1")
    .inLibrary("com.typesafe" % "config" % "1.3.0")
    .inProject
)
```

**Comment by sfc-gh-pbennes** [comment]
Taking a look at the shaded `net.snowflake:snowflake-ingest-sdk:1.1.3`, the jar contains unrelocated classes of its dependencies which would conflict with any project that includes those dependencies through any build tool normal process. So this is a valid issue for the shaded jar. We shouldn't be including unrelocated classes that can conflict.

```sh
curl -LO https://repo1.maven.org/maven2/net/snowflake/snowflake-ingest-sdk/1.1.3/snowflake-ingest-sdk-1.1.3.jar
unzip -l snowflake-ingest-sdk-1.1.3.jar | egrep -v 'net/snowflake/ingest/' | fgrep -c '.class'
30629
unzip -l snowflake-ingest-sdk-1.1.3.jar | egrep -v 'net/snowflake/ingest/' | fgrep '.class' | head
      724  11-05-2022 20:37   META-INF/versions/9/module-info.class
     2208  12-03-2021 04:42   com/github/benmanes/caffeine/SCQHeader$HeadAndTailRef.class
     1040  12-03-2021 04:42   com/github/benmanes/caffeine/SCQHeader$HeadRef.class
     2406  12-03-2021 04:42   com/github/benmanes/caffeine/SCQHeader$PadHead.class
     2498  12-03-2021 04:42   com/github/benmanes/caffeine/SCQHeader$PadHeadAndTail.class
      651  12-03-2021 04:42   com/github/benmanes/caffeine/SCQHeader.class
     2523  12-03-2021 04:42   com/github/benmanes/caffeine/SingleConsumerQueue$1.class
     1298  12-03-2021 04:42   com/github/benmanes/caffeine/SingleConsumerQueue$LinearizableNode.class
     2541  12-03-2021 04:42   com/github/benmanes/caffeine/SingleConsumerQueue$Node.class
     1839  12-03-2021 04:42   com/github/benmanes/caffeine/SingleConsumerQueue$SerializationProxy.class
```

> We have faced the issue too. Our solution was to change the order of loaded jars, to make snowflake-sdk the last one.

Maven's classpath ordering is based on the order dependencies appear in pom.xml, so moving the SDK to last position lowers the priority of classes that are part of this shaded jar.

It's not clear to me either why shading was introduced as part of #19. We're actively trying to address build related issues in the current shaded jar as well as start to produce an unshaded jar which I'm trying to direct people towards. Hopefully this issue will be addressed soon.

A question for @jdcaperon and @leszekgruchala do you have any preference one way or the other for a shaded or unshaded jar? Would it be problematic if we stopped producing a shaded jar?

**Comment by sfc-gh-pbennes** [comment]
Yeah I think the shaded jar was the only thing being published for a while. I'm new to being a contributor to this library and the related snowflake-jdbc library. I've spent some time as a Java build engineer internally at Snowflake on other projects and I'm trying to help improve the build for this and snowflake-jdbc. I've been recommending that if at all possible we move away from shading so it's good to hear about the switching to unshaded jar.

**Comment by sfc-gh-pbennes** [comment]
Yeah hopefully most if not all of the shading being done currently isn't necessary. There _might_ be some cases where it's convenient, but as someone that _consumes_ a lot of dependencies I much prefer to be making the choice on what and when to shade if necessary. I strongly prefer that libraries don't shade, or at the very least if they do it's an option and not an unavoidable default. Having a shaded library with old unrelocated classes of widely used dependencies is particularly not optimal.

**Comment by leszekgruchala** [comment]
I see `1.1.3-unshaded` works for us. If shading is unnecessary, then I think making at some point a new release simply without shading would make it simpler for everyone.

**Comment by leszekgruchala** [comment]
Hey, I am looking forward for the release of [this fix](https://github.com/snowflakedb/snowflake-ingest-java/pull/492) with a shaded library again (1.1.4?), as currently the unshaded one brings a lot of dependencies where some of them have low, mid, and high vulnerabilities.

**Comment by sfc-gh-pbennes** [comment]
> Hey, I am looking forward for the release of [this fix](https://github.com/snowflakedb/snowflake-ingest-java/pull/492) with a shaded library again (1.1.4?), as currently the unshaded one brings a lot of dependencies where some of them have low, mid, and high vulnerabilities.

If you're using Maven you can control the version of these dependencies with a `<dependenciesManagement>` section if there are patched versions of these dependencies available. Do you have an example of one of these dependencies offhand and I can give you a specific example of what that would look like? Also it's worth noting that shading a vulnerable dependency isn't really an improvement on the situation because a vulnerable dependency is still present it just makes it harder for security tools to detect.

Here's a general example of a dependencyManagement section, used to control the version of particular dependencies anywhere they appear in the dependency graph, including transitive dependencies (dependencies of dependencies): https://github.com/snowflakedb/snowflake-ingest-java/blob/master/pom.xml#L68

**Comment by leszekgruchala** [comment]
@sfc-gh-pbennes thanks for the insights, but working over the `dependenciesManagement` is not a viable, long term solution.

The snowflake SDK provides **a lot** of nested dependencies my production code is not interested in. Without shading, all those dependencies are listed as "mine" even though I don't use them at all. But, the security tools don't know that and treat them as used, indicating about any potential security vulnerabilities.

For example, `org.eclipse.jetty:jetty-server:9.4.48` has [this vulnerability](https://github.com/eclipse/jetty.project/security/advisories/GHSA-p26g-97m4-6q7c). And this shows how deeply nested it is:

```
net.snowflake:snowflake-ingest-sdk:1.1.3-unshaded
 + org.apache.arrow:arrow-vector:10.0.0
   + org.apache.hadoop:hadoop-common:3.3.5
     + org.apache.avro:avro:1.7.7
       + org.eclipse.jetty:jetty-server:9.4.48.v20220622
```

To solve this, I would have to analyze and control the version of many nested dependencies. Not only it is risky as I may introduce some incorrect version for the Snowflake SDK with an issue seen on runtime, but also this is something I need to revisit with every Snowflake SDK update.


With proper shading, you use what you want (with my hope that you maintain latest possible versions alone and security checks on the SDK) and my application is not affected, as long as I don't use some dependency directly.

**Comment by sfc-gh-pbennes** [comment]
Possibly the short term fix for some of these dependencies is to see if we can exclude these dependencies from the graph. I doubt we need either avro or jetty-server. [Actually it looks like we try to do this but the exclusions may be in the wrong place, dependencyManagement vs dependencies](https://github.com/snowflakedb/snowflake-ingest-java/blob/master/pom.xml#L147-L167).

> Without shading, all those dependencies are listed as "mine" even though I don't use them at all. But, the security tools don't know that and treat them as used, indicating about any potential security vulnerabilities.

From a security point of view for Java this is correct. While application code might not reference the vulnerable code it may still be exploitable just by being present, depending on the nature of the exploit!

As far as Java is concerned a class is either present on the classpath or it isn't. The classpath doesn't distinguish between first party classes and third party classes from libraries/dependencies. Hierarchichal packages are a useful mental model for us humans but at runtime all classes get added to the classpath which is basically just a flat list. This is why [checking for duplicate classes is good practice](https://www.mojohaus.org/extra-enforcer-rules/banDuplicateClasses.html) because the first version of a class on the classpath is the one that gets loaded and used. I think most security tools operate on library maven coordinates detected in published pom.xmls. If a security tool finds problematic dependencies in an unshaded library but doesn't in a shaded library that's a false negative. The vulnerable code hasn't been made not-vulnerable by relocation, though maybe less exploitable depending on the nature of the vulnerability. This is one of the reasons I advocate against shading. https://jlbp.dev/JLBP-18 covers other good reasons.

> To solve this, I would have to analyze and control the version of many nested dependencies.

Correct. This is what I recommend and what we do. [dependencyManagement section](https://github.com/snowflakedb/snowflake-ingest-java/blob/master/pom.xml#L68) is what we do to manage dependencies and keep them up to date. Security patching is one reason and upper version alignment https://jlbp.dev/JLBP-16 is another. The [Linkage Checker](https://github.com/GoogleCloudPlatform/cloud-opensource-java/wiki/Linkage-Checker-Enforcer-Rule) is how we check library compatibility when upgrading a library with dependencyManagement.

We try to do many of the best practices listed at https://jlbp.dev/ and I similarly recommend them for anyone working with Java.

**Comment by leszekgruchala** [comment]
Very good points. Indeed, the fact I don't see those nested deps doesn't mean they are not there and vulnerable, yet, a bit harder to use.

We also maintain some of the versions of the nested dependencies, but the amount of them from snowflake-sdk is quite big.

I'd like to raise a question, do you really need all of them? If it's possible to reduce them, it's easier for the clients to handle the versions too.

Additionally, do you run some vulnerability checks tool(s) too?

**Comment by sfc-gh-lsembera** [comment]
We just released [SDK 2.0.0](https://central.sonatype.com/artifact/net.snowflake/snowflake-ingest-sdk/2.0.0/versions) with cleaned up dependencies and other improvements related to shading, please give it a try. Unshaded version is going to be released soon.
